### PR TITLE
Add method to set/get cpu frequency governor

### PR DESCRIPTION
**Pull request by sathnaga**
Add method to change given cpu frequency governor systemwide
and choose a random available one if special governor value
`random` is given.

Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

File: avocado/utils/cpu.py [hunk 1]
```diff
@@ -281,3 +282,49 @@ def set_cpuidle_state(state_number="all", disable=True, setstate=None):
...
+                     option to choose random profile among available ones.
+    """
+    avl_gov_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_available_governors"
+    cur_gov_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
+    if os.access(avl_gov_file, os.R_OK) and os.access(cur_gov_file, os.W_OK):
+        avl_govs = open(avl_gov_file, 'r').read().strip().split(' ')
+        cur_gov = open(cur_gov_file, 'r').read().strip()
+        if governor == "random":
+            avl_govs.remove(cur_gov)
+            governor = avl_govs[random.randint(0, len(avl_govs)-1)]
```
Comment (ldoktor) [inline avocado/utils/cpu.py]: wouldn't `random.choice(avl_govs)` fit better?

File: avocado/utils/cpu.py [hunk 2]
```diff
@@ -281,3 +282,49 @@ def set_cpuidle_state(state_number="all", disable=True, setstate=None):
...
+        cpus_list = range(total_cpus_count())
+        if governor not in avl_govs:
+            logging.warning("Trying to change unknown frequency "
+                            "governor: %s", governor)
+        for cpu in cpus_list:
+            cur_gov_file = "/sys/devices/system/cpu/cpu%s/cpufreq/scaling_governor" % cpu
+            with open(cur_gov_file, 'w') as fl:
+                try:
+                    fl.write(governor)
+                    fl.close()
```
Comment (ldoktor) [inline avocado/utils/cpu.py]: no need to close the file here, `with` does that on `__exit__`.

File: avocado/utils/cpu.py [hunk 3]
```diff
@@ -281,3 +282,49 @@ def set_cpuidle_state(state_number="all", disable=True, setstate=None):
...
+        logging.error("Could not locate frequency governor sysfs entries")
+        return False
+
+
+def get_cpufreq_governor():
+    """
+    Get current cpu frequency governor
+    """
+    cur_gov_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
+    if os.access(cur_gov_file, os.W_OK):
```
Comment (ldoktor) [inline avocado/utils/cpu.py]: This should be `os.R_OK`

File: avocado/utils/cpu.py [hunk 4]
```diff
@@ -281,3 +282,49 @@ def set_cpuidle_state(state_number="all", disable=True, setstate=None):
...
+
+def set_cpufreq_governor(governor="random"):
+    """
+    To change the given cpu frequency governor
+
+    :param governor: frequency governor profile name whereas `random` is default
+                     option to choose random profile among available ones.
+    """
+    avl_gov_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_available_governors"
+    cur_gov_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
```
Comment (ldoktor) [inline avocado/utils/cpu.py]: How about this:

```python
cur_gov = get_cpufreq_governor()
if not cur_gov:
    logging.warning("Failed to get current frequency governor.")
    return False
if not (os.access(avl_gov_file, os.R_OK) and os.access(cur_gov_file, os.W_OK)):
    logging.warning("Access to files required to change frequency governor failed."
    return False
# and then the full body without the unnecessary `else`
```

File: avocado/utils/cpu.py [hunk 5]
```diff
@@ -281,3 +282,49 @@ def set_cpuidle_state(state_number="all", disable=True, setstate=None):
...
+        return False
+
+
+def get_cpufreq_governor():
+    """
+    Get current cpu frequency governor
+    """
+    cur_gov_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
+    if os.access(cur_gov_file, os.W_OK):
+        return open(cur_gov_file, 'r').read().strip()
```
Comment (ldoktor) [inline avocado/utils/cpu.py]: Please use `with open...` otherwise we'll get "unclosed file errors" on py3.

File: avocado/utils/cpu.py (hunk 1 above)
Comment (sathnaga) [reply avocado/utils/cpu.py]: Sure, it is, thanks :-)

File: avocado/utils/cpu.py (hunk 2 above)
Comment (sathnaga) [reply avocado/utils/cpu.py]: Sure

**Comment by sathnaga** [comment]
@ldoktor Thanks for the detailed review and suggestions, regarding per cpu granularity, this method won't suite as we want to have system wide change to choose a different governors.

Will think about the usecase and add it in case in future.

**Comment by sathnaga** [comment]
@ldoktor have addressed your review comments, pls help in merging, thanks in advance.

**Comment by ldoktor** [comment]
Oh, the base of this is really old, but it only adds new API so it should not break in between. Let me run full make-check instead of checking all things in between...

**Pull request merged**